File: src/TrainingLoadAnalyzer.Web/Features/Dashboard/Display.cs
using System.Globalization;

namespace TrainingLoadAnalyzer.Web.Features.Dashboard;

/// <summary>
///   How a figure reaches the page (C99, C100).
/// </summary>
/// <remarks>
///   <para>
///     Everything numeric goes through here, against <see cref="CultureInfo.InvariantCulture"/>
///     rather than the server's. SC-002 writes its worked example as "45.3", and on a machine set
///     to Finnish the obvious rendering produces "45,3" — and, for a negative form, "−13,2" with a
///     U+2212 minus sign rather than a hyphen (research R9).
///   </para>
///   <para>
///     One place, so there is one thing to get right. The same reasoning governs
///     <see cref="MetricsChart"/>, where the consequence is worse than an odd-looking number: a
///     decimal comma inside an SVG coordinate list silently changes the geometry.
///   </para>
/// </remarks>
public static class Display
{
    /// <summary>The em dash a missing figure shows, never a zero (US1 scenario 2, C100).</summary>
    public const string Missing = "—";

    private const string OneDecimal = "0.0";

    /// <summary>Fitness, fatigue or form, to one decimal place (FR-001 - FR-003, SC-002).</summary>
    public static string Metric(double? value) =>
        value is null ? Missing : value.Value.ToString(OneDecimal, CultureInfo.InvariantCulture);

    /// <summary>A training load in TRIMP points (FR-004, FR-007).</summary>
    public static string Points(decimal? value) =>
        value is null ? Missing : value.Value.ToString(OneDecimal, CultureInfo.InvariantCulture);

    /// <summary>A whole-number percentage, sign included (FR-005).</summary>
    public static string Percent(decimal? fraction) =>
        fraction is null
            ? Missing
            : (fraction.Value * 100).ToString("+0;-0;0", CultureInfo.InvariantCulture) + "%";

    /// <summary>A duration as hours and minutes, or minutes alone under an hour (FR-007).</summary>
    public static string Duration(TimeSpan movingTime) =>
        movingTime.TotalHours >= 1
            ? string.Create(CultureInfo.InvariantCulture, $"{(int)movingTime.TotalHours}h {movingTime.Minutes:00}m")
            : string.Create(CultureInfo.InvariantCulture, $"{movingTime.Minutes}m");

    /// <summary>A calendar day, unambiguous on any machine and in any locale.</summary>
    public static string Day(DateOnly day) => day.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
}

File: alt-stack/backend/src/tla/dashboard/display.py
"""How a figure reaches the page: .NET's Display, reproduced without the ambient locale (research R2).

.NET formats a double to 15 significant digits and then rounds half away from zero; Python's own
float formatting rounds the exact binary value half to even, and disagrees on 5 of 13 probes. So a
metric goes through its 15-digit text and a Decimal quantize. No `locale`, no f-string float format.
"""

from datetime import date, timedelta
from decimal import ROUND_HALF_UP, Context, Decimal

MISSING = "—"

_ONE_DECIMAL = Decimal("0.1")
_WHOLE = Decimal("1")
_HUNDRED = Decimal(100)

# Wide enough that quantizing any figure this page can show never runs out of digits.
_CONTEXT = Context(prec=60)


def metric(value: float | None) -> str:
    """Fitness, Fatigue or Form, to one decimal place. A negative that rounds to zero keeps its sign."""
    if value is None:
        return MISSING
    rounded = Decimal(format(value, ".15g")).quantize(_ONE_DECIMAL, rounding=ROUND_HALF_UP, context=_CONTEXT)
    return format(rounded, "f")


def points(value: Decimal | None) -> str:
    """A training load. A decimal that rounds to zero has no sign, as .NET's decimal prints it."""
    if value is None:
        return MISSING
    rounded = value.quantize(_ONE_DECIMAL, rounding=ROUND_HALF_UP, context=_CONTEXT)
    return format(rounded.copy_abs() if rounded == 0 else rounded, "f")


def percent(fraction: Decimal | None) -> str:
    """A whole-number percentage: "+" above zero, "-" below, and "0" when it rounds to zero."""
    if fraction is None:
        return MISSING
    whole = _CONTEXT.multiply(fraction, _HUNDRED).quantize(_WHOLE, rounding=ROUND_HALF_UP, context=_CONTEXT)
    if whole > 0:
        return f"+{format(whole, 'f')}%"
    if whole < 0:
        return f"{format(whole, 'f')}%"
    return "0%"


def week_change(change: Decimal) -> str:
    """MetricRow's caption: a "+" for a change of zero or more, then the decimal."""
    return ("+" if change >= 0 else "") + points(change)


def duration(moving_time: timedelta) -> str:
    """Hours and padded minutes, or minutes alone under an hour. Seconds are dropped, never rounded."""
    total_minutes = moving_time // timedelta(minutes=1)
    hours, minutes = divmod(total_minutes, 60)
    return f"{hours}h {minutes:02d}m" if hours >= 1 else f"{minutes}m"


def day(value: date) -> str:
    return value.isoformat()
